### ISSUE TITLE
卷积的TIMG2COL和weight TLOAD实现存在BUG

### ISSUE THREAD

**Issue by CCxx00** [LinxISA/Linx-TileOP-API#195]
[conv2d.tar.gz](https://github.com/user-attachments/files/32456196/conv2d.tar.gz)

**验证环境**

| 组件 | 版本 |
|---|---|
| llvm-project | `ccfa8cd9ea9b`（dev-llvm15_56 头，含 `69d6e6d8`） |
| Linx-TileOP-API | `3be8652`（linxisa-v0.58.0-194；问题 1/2/3 已确认 head `a7135b7` 仍存在） |
| SuperScalarModel (gfrun) | `0c6dd7b4`（fix/gfrun-717-pto291-subview-ordering；问题 4 已确认 head `3ca45178` 与 main 仍存在） |
| 编译选项 | `-mlxbc -fenable-matrix -O2 -mllvm -enable-all-vector-as-tilereg=true -std=c++20` |

---

## Issue 1（Linx-TileOP-API）：weight TLOAD 封装 B.DATR 文本拼写被汇编器拒绝

**Title**: weight-TLOAD wrapper emits assembler-rejected `B.DATR layout<N>` spelling; numeric layout code required

### 复现

```cpp
#include <common/pto_tileop.hpp>
using namespace pto;

using WeightOut = Tile<Location::Right, float, 16, 32, BLayout::RowMajor, 16, 32>;
using WeightShared = SharedTile<WeightOut>;
using WeightGM = global_tensor<float, RowMajor<16, 32>>;

__attribute__((noinline)) void weight_tload(WeightShared &dst, const WeightGM &src) {
  // 与 docs/tileop-usage/tlsu/load-store-move/TLOAD.md 及
  // test/tileop_api/src/SharedTLoad.cpp weight_tload() 相同的调用形式
  TLOAD<OIHW2NK>(dst, src, make_weight_tload_params(16, 16, 1, 1));
}
```

```bash
clang++ -c -mlxbc -fenable-matrix -O2 \
    -mllvm -enable-all-vector-as-tilereg=true -std=c++20 \
    -I <Linx-TileOP-API>/include repro.cpp -o repro.o
```

### 实际行为

```
include/jcore/template_asm.hpp:2886:8: error: Match Instruction Error!
      "B.DATR layout%c[WeightLayout], DTYPE_NONE, Zero\n"
<inline asm>:2:1: note: instantiated into assembly here
B.DATR layout11, DTYPE_NONE, Zero
```

（`WeightLayout=OIHW2NK=11`，`layout` 文本前缀 + 数字的拼写不被接受；编译退出码 74。）

### 期望行为

llvm dev-llvm15_56 `f81c7f5270a4`（"[LinxV5] Add B.DATR Layout+DataType+PadValue(+RMode) aliases (issue #100)"）已接受**纯数字布局码**形式；手工验证 `B.DATR 11, DTYPE_NONE, Zero` 可完整汇编（其余 bundle 不变时 rc=0）。封装应发射数字码。

### 根因

`include/jcore/template_asm.hpp`（3be8652:2886，head a7135b7 同）：

```cpp
"B.DATR layout%c[WeightLayout], DTYPE_NONE, Zero\n"
```

`layout` 前缀文本形式从未被 MC 接受；对照 TIMG2COL 封装（14050）的 `B.DATR %c[Layout], DTYPE_NONE, Zero`（数字码）即可。

### 修复建议（本地已验证）

```diff
-      "B.DATR layout%c[WeightLayout], DTYPE_NONE, Zero\n"
+      "B.DATR %c[WeightLayout], DTYPE_NONE, Zero\n"
```

依赖：需 llvm ≥ `f81c7f5270a4`。另注：在更早的 llvm（73cbdf34fbd0，即 < abeccf4937e8）上该封装还会先触发 Simple Register Coalescing `joinCopy → setRegClass "Invalid RC"` abort——该 llvm 侧问题已由 `abeccf4937e8` 修复，但 TileOP 拼写问题独立存在。

---

## Issue 2（Linx-TileOP-API）：weight TLOAD 封装参数缺 zero 折叠防护，常量 0 StartWord 导致 B.IOR 丢源

**Title**: weight-TLOAD wrapper loses the StartGPR B.IOR source when ShapeWord/StartWord fold to the zero register

### 复现

上述 Issue 1 修复后，`n_start == 0 && k_start == 0`（首个权重组，或循环全展开后 j=0 迭代）时编译通过，但 gfrun 运行期断言：

```
gfrun: illegal instruction: ASSERTION FAILED:
ExecuteSharedWeightTLOAD(block) && "PTO #266 Shared weight TLOAD schema or memory footprint is illegal"
, func ExecuteSharedTMA, file SoftCore.cpp:3062
```

gfrun trace 显示 bundle 的 B.IOR 只剩 2 个源（应为 3 个：GMBase, ShapeGPR, StartGPR）：

```
M9216  B.IOS [0x0.d], [0x1.d], [0x5.d]        ; slot 0, PE0, SizeCode 5 — 正常
M9217  B.IOR a2, a7                           ; 仅 2 源！a7 = 0x20200200008 = ShapeWord
```

### 根因

封装将 `params.shape_word` / `params.start_word` 直接作为 `"r"` 内联 asm 输入（template_asm.hpp:2901-2902）。`start_word` 为编译期常量 0 时，编译器将其折叠进 zero 寄存器；zero 寄存器源在 B.IOR 编码中被丢弃，bundle 违反 PTO-BSTART-TLOAD-WEIGHT-NK-CONTRACT-001 的三源 schema。

TIMG2COL 封装对三个参数字已有 `asm("" : "+r"(paramN))` opaque 防护（template_asm.hpp:14043-14048），weight 封装遗漏了同样的处理。

### 修复建议（本地已验证）

```diff
+  uint64_t shape_word = params.shape_word;
+  asm("" : "+r"(shape_word));
+  uint64_t start_word = params.start_word;
+  asm("" : "+r"(start_word));
   const size_t valid_k = dst.GetValidCol();
   const size_t valid_n = dst.GetValidRow();
   asm volatile(
       ...
-      : [Base] "r"(src.data()), [Shape] "r"(params.shape_word),
-        [Start] "r"(params.start_word), [PEMask] "i"(PEMask),
+      : [Base] "r"(src.data()), [Shape] "r"(shape_word),
+        [Start] "r"(start_word), [PEMask] "i"(PEMask),
```

---

## Issue 3（Linx-TileOP-API）：TIMG2COL 封装无法与协作 TMATMUL 组合（缺 groupM 显式参数重载）

**Title**: TIMG2COL wrapper cannot compose with the cooperative Local-A TMATMUL: LB1 derives from the dst type's ValidRow instead of an explicit group-row argument

### 问题

TIMG2COL 封装只支持 Local Left CubeM16/M32 目的端，LB1（4-PE 组总行数）取自目的端 Tile 类型的 `ValidRow`（立即数）。这使它与协作 TMATMUL 的 Local-A/Shared-B 形式**在任何组尺寸下都无法组合**：

1. **多 PE 组（64 行 @ 16 行/PE，或 128 行 @ 32 行/PE）静态契约冲突**：
   - TIMG2COL 要求 dst 类型 `ValidRow == 组总行数`（64/128），且 `Tile` 强制 `ValidRow <= Rows`（pto_tile.hpp:1030）；
   - 协作 TMATMUL 要求 `A::ValidRow == 每 PE 分片（16/32）`（template_asm.hpp `validate_matrix_contract`：`static_assert(AValidRows == 16 || AValidRows == 32, "Cooperative Local A per-PE shard must be 16 or 32 rows ...")`）；
   - 同一 Tile 对象无法同时满足两者。

2. **单 PE 块组（group_M == pe_m，16/32 行）运行期崩溃**：类型检查可通过，但 4-PE SPMD 下其余 PE 是 0 行 PE——按 ASL 契约 0 行 PE **不分配 Local 目的端**（gfrun AccumulateBlockInfo.cpp:2092-2102：`zeroRowTimg2Col → inst->dsts.clear()`），而协作 TMATMUL 在**每个 PE** 上都通过 per-PE tile 队列解析 A 操作数（AccumulateBlockInfo.cpp:1989 `tileReg.LookUp`），导致：

```
ASSERTION FAILED: false
FuncTBackToFrontIndex: out of index range, func BackToFrontIndex, file common/DataStruct/BaseROB.h:101
（调用栈：BaseROB<TileEntry>::BackToFrontIndex ← SoftCore::AccumulateBlockInfo）
```

### 修复建议（本地扩展已实现并端到端验证）

镜像 TMATMUL 的 groupM 重载（`TMATMUL(c, a, b, options, groupM)`），为 TIMG2COL 增加运行时 groupRows 参数重载：LB1 由 `"r"` GPR 携带，目的端保持每 PE 分片类型（16/32 行），使全部 PE 物化自己的分片：

```cpp
template <is_tile_data_v tile_shape_out, is_global_data_v gm_shape>
  requires(tile_shape_out::Loc == Location::Left &&
           (tile_shape_out::BFractal == BLayout::CubeM16 ||
            tile_shape_out::BFractal == BLayout::CubeM32))
void TIMG2COL(tile_shape_out &dst, gm_shape &src,
              uint64_t param0, uint64_t param1, uint64_t param2,
              size_t groupRows) {
  // （dtype 静态约束同基础封装）
  asm volatile(
      "BSTART.TIMG2COL %D[DataType]\n"
      "B.DATR %c[Layout], DTYPE_NONE, Zero\n"
      "B.DIM zero, %c[ValidCol], ->lb0\n"
      "B.DIM %[GroupRows], 0, ->lb1\n"      // 运行时组总行数
      "B.DIM zero, %c[TotalCol], ->lb2\n"
      "B.IOR [%[GMBase], zero, zero], []\n"
      "B.IOR [%[Param0], %[Param1], %[Param2]], []\n"
      "B.IOT mask=1111, last, ->%[Dst]<%Z[TileSize]>\n"
      : [Dst] "=Tr"(dst.data())
      : [GMBase] "r"(src.data()),
        [DataType] "i"(type_traits<typename gm_shape::DType>::TypeCode),
        [Layout] "i"(tile_shape_out::BFractal == BLayout::CubeM16 ?
                       LayoutCvtEnum::ND2M16 : LayoutCvtEnum::ND2M32),
        [ValidCol] "i"(tile_shape_out::ValidCol),
        [TotalCol] "i"(tile_shape_out::Cols),
        [Param0] "r"(param0), [Param1] "r"(param1), [Param2] "r"(param2),
        [GroupRows] "r"(groupRows),
        [TileSize] "i"(tile_type_traits<typename tile_shape_out::TileDType>::TilesizeCode)
      : "memory");
}
```

使用方式（与 gfrun 模型语义逐项对齐验证）：

```cpp
using tileAShard = CubeTileM16<float, 16, K>;   // 每 PE 分片类型
tileAShard tA;                                   // 所有 PE 执行
TIMG2COL(tA, gA, p0, p1, p2, /*groupRows=*/64);  // lb1=64 → 4 PE 各物化 16 行
TMATMUL(tC, tA, tW, fixp::keep_acc(), /*groupM=*/64);
```

gfrun 侧无需改动：`Timg2ColValidRowsForLayout(ND2M16, 64, pe)` 行切片、per-PE `tileInfo->validRow = peRows = 16`、TMATMUL 协作 `localRows = CooperativeLocalM(64) = 16` 三者天然对齐。

---

## Issue 4（SuperScalarModel/gfrun）：`movr` 写 zero 寄存器目的端未建模（Calculate error）

**Title**: gfrun rejects `movr <zero> <- <gpr>` (write to the zero register): SetDstOperand drops dst==0 and CalcInstMove fails

### 复现

llvm dev-llvm15_56 ≥ `28e88c9049f6`（"[LinxV5] Never absorb scalar body into Tile blocks (A3)"；验证于 `ccfa8cd9ea9b`）会在 Tile 块边界发射独立 STD 块的写 zero 空操作，例如（实测 ELF，TIMG2COL 块之后）：

```
123c2: B.IOT mask=1111, last, ->t<2KB>    ; TIMG2COL 块结束
123c6: C.BSTART.STD
123c8: 0206  c.movr a6, ->zero            ; dst=zero(bits[15:11]=0), src=a6(bits[10:6]=8)
123ca: BSTART.CUBE TMATMUL ...
```

gfrun 运行期：

```
gfrun: illegal instruction: ASSERTION FAILED: false
Calculate error:
M10693 |TPC:0x123c8 ... |a6:0x0 | | |c.movr a6 |bin: 0x206
, func Calculate, file SoftCore.cpp:962
```

### 根因

1. `MInst::SetDstOperand`（isa/MInst.cpp:1749）：`STORE_TO_REG_DST` 分支只接受 r1-r23 / U 队列 / T 队列，**dst==0 无分支**，操作数保持 `OPD_INVALID`；
2. 调用处 `if (dst0->type != OperandType::OPD_INVALID) dsts.push_back(dst0);`（isa/MInst.cpp:1126）将其丢弃 → `dsts.size()==0`；
3. `CalcInstMove`（isa/calculate/others/Others.cpp:7-14）要求 `srcs.size()==1 && dsts.size()==1` → 返回 false → `ASSERT(false) << "Calculate error"`。

而模型基础设施本就支持该语义，仅缺解码分支：
- 源读取：`GetSrcOperandData` 中 `OPD_ZERO → data = 0`（emulator/engine/OperandIO.cpp:70-72，恒 0，不读寄存器堆）；
- 目的写回：`SetDstOperandData` 中 `case OPD_ZERO: // Readonly; break;`（OperandIO.cpp:245-246，**写 zero 已按丢弃处理**）；
- `CTGenEngine` 已在多处使用 `OPD_ZERO` 目的端占位（CTGenEngine.cpp:246/517/529）。

### 修复建议（本地已验证，+2 行）

```diff
 void MInst::SetDstOperand(Operand& operand, int64_t dst, int32_t dstType)
 {
     if (dstType == STORE_TO_REG_DST) {
         if (dst >= DST_GREG_LEFT_BOUND && dst <= DST_GREG_RIGHT_BOUND) {                    // r1-r23
             operand = Operand(OperandType::OPD_GREG, static_cast<uint64_t>(dst));
+        } else if (dst == 0) {                                                              // r0: discarding destination
+            operand = Operand(OperandType::OPD_ZERO, 0);
         } else if (dst == DST_UREG_BOUND) {                                                 // U队列, value无意义
```

已确认 fix 分支头（`3ca45178`）与 main 的 `SetDstOperand` 均无 dst==0 处理。

### 验证

打补丁后，依赖该 llvm 版本编译的 conv2d 封装化内核（TIMG2COL + weight TLOAD + 协作 TMATMUL）在 gfrun 4-PE 下完整运行，16384 元素 golden 全匹配（max_abs_diff=2.682e-07）。

---

## 附：端到端验证载体

SuperNPUBench `benchmark/one-level-arch/kernels/basic_op/conv2d/conv2d_img2col.hpp`（封装化版本）同时依赖 Issue 1-4 的修复；四项本地补丁齐备时：

```
$ python3 golden_conv2d_img2col.py --elf .../kernel_conv2d_conv2d_img2col_PE4.elf
elements=16384  max_abs_diff=2.682e-07  mismatches=0
PASS: conv2d_img2col matches torch.nn.functional.conv2d
```

**Comment by VV0003** [comment]
## 判定：四项全部属实；TileOP 侧两项已修复（PR #196），Item 1 此前已修，Item 4 属模型侧

感谢极其完整的报告——四项的根因分析经独立核实全部正确。逐项状态：

### Item 1（weight TLOAD `layout<N>` 拼写）：已由 PR #188 修复

`linx ≥ 83bfb85` 的封装已发射 ISA 名 `OHWI2NK`/`OIHW2NK`（比 issue 建议的数字码更强的形式，手工验证 `B.DATR 11, ...` 与 `B.DATR OIHW2NK, ...` 均可汇编）。报告用的 `3be8652` 早于该修复。llvm ≥ `abeccf49` 的 coalescer abort 也已修（那正是 #99）。

### Item 2（StartWord 折叠进 zero 丢 B.IOR 第三源）：属实，PR #196 修复

本地在 `533ad1b` 复现：`make_weight_tload_params(16,16,1,1)`（start=0）发出 `B.IOR [a3, a4], []` 仅两源。根因与报告一致：`"r"` 输入的编译期常量 0 被折叠进 zero 寄存器，B.IOR 编码丢弃 zero 源，违反 `PTO-BSTART-TLOAD-WEIGHT-NK-CONTRACT-001` 的 "exactly one three-source B.IOR binds GMBase, ShapeGPR, StartGPR"。修复采用 issue 建议的 opaque 防护（TIMG2COL 同款 `asm("" : "+r"(x))`）：

```text
修复前:  B.IOR [a3, a4], []          ← 2 源
修复后:  B.IOR [a1, a4, a5], []      ← GMBase, ShapeGPR, StartGPR
```

### Item 3（TIMG2COL 与协作 TMATMUL 无法组合）：属实，PR #196 新增 groupRows 重载

ASL 独立核实了报告的分析：`BundleTIMG2COLValidRowForOutput`（timg2col-schema.asl）定义 `M_per_PE = 16/32`、LB1 为组总行数（1..128）、零行 PE "remain collective participants but perform no allocation"——与 `validate_matrix_contract` 的 `A::ValidRow == 16/32` 分片契约确实无法用同一个 Tile 类型同时满足。修复按 issue 给出的方案：新增运行时 `groupRows` 参数重载（LB1 走 GPR，目的端保持每 PE 分片类型），镜像 TMATMUL 的 groupM 重载。实测：

- 常量 `groupRows=64` → 折叠为合法立即数 `C.B.DIMI 64, ->lb1`；
- 运行期 `groupRows` → 保持寄存器形式 `B.DIM a2, 0, ->lb1`；
- `TIMG2COL(分片, GM, p0..p2, groupRows)` + `TLOAD<OIHW2NK>` + `TMATMUL(c, a, w, keep_acc, groupM)` 端到端编译通过。

### Item 4（gfrun `movr -> zero` 未建模）：核实属实，属 SuperScalarModel 侧

报告的 `SetDstOperand` dst==0 缺分支分析与修复建议（+2 行映射 `OPD_ZERO`）我们核实了模型侧代码路径（`OperandIO.cpp` 的读写两端都已支持 OPD_ZERO，仅解码缺分支）成立。按分工模型仓不在本仓修复范围，建议在 LinxISA/SuperScalarModel 开独立 issue 并引用本报告的 diff；该问题只在 llvm ≥ `28e88c90`（A3 块边界写 zero 空操作）后出现。

### 验证与工具链要求

- PR #196（本仓）：新增注册测试 `Issue195ConvCompose.cpp` 覆盖三个场景；`make check`、`test_v058_engine_contract` 全绿；
- 完整 conv2d 端到端（含 golden 比对）需要：llvm ≥ `ccfa8cd9`（含 A3）+ 本 PR + 模型侧 Item 4 补丁，与报告的端到端验证矩阵一致。
